wgpu-hal/src/lib.rs

```diff
@@ -1242,8 +1242,38 @@ pub trait CommandEncoder: WasmNotSendSync + fmt::Debug {
 
     // pass common
 
-    /// Sets the bind group at `index` to `group`, assuming the layout
-    /// of all the preceding groups to be taken from `layout`.
+    /// Sets the bind group at `index` to `group`.
+    ///
+    /// If this is not the first call to `set_bind_group` within the current
+    /// render or compute pass:
+    ///
+    /// - If `layout` contains `n` bind group layouts, then any previously set
+    ///   bind groups at indices `n` or higher are cleared.
+    ///
+    /// - If the first `m` bind group layouts of `layout` are equal to those of
+    ///   the previously passed layout, but no more, then any previously set
+    ///   bind groups at indices `m` or higher are cleared.
+    ///
+    /// It follows from the above that passing the same layout as before doesn't
+    /// clear any bind groups.
+    ///
+    /// # Safety
+    ///
+    /// - `index` must be the valid index of some bind group layout in `layout`.
+    ///   Call this the "relevant bind group layout".
+    ///
+    /// - The layout of `group` must be equal to the relevant bind group layout.
+    ///
+    /// - The length of `dynamic_offsets` must match the number of buffer
+    ///   bindings [with dynamic offsets][hdo] in the relevant bind group
+    ///   layout.
+    ///
+    /// - If those buffer bindings are ordered by increasing [`binding` number]
+    ///   and paired with elements from `dynamic_offsets`, then each offset must
+    ///   be a valid offset for the binding's corresponding buffer in `group`.
+    ///
```

Comment:
suggestion: Because we're talking about using this in the context of a pass (`within the current render or compute pass`), it seems like we might be missing an explicit safety precondition here for "needs to have begun (but not ended) a render or compute pass". Thoughts?

Reply:
suggestion: If you agree with ☝🏻 the above suggestion, we might amend the summary line of the method to be:

```diff
-    /// Sets the bind group at `index` to `group`.
+    /// Sets the bind group of the current render or compute pass at `index` to `group`.
```

Reply:
> we might be missing an explicit safety precondition

Oh, we definitely are. I have some other commits in progress that will extend the "states" described in the `CommandEncoder` documentation to include "recording render pass" and "recording compute pass". But I'd like to leave that for a separate PR, because it touches pretty much all the `CommandEncoder` methods if done right.

For now, I like your proposed wording.